nx/lib/nx.ex

```diff
@@ -7142,6 +7142,61 @@ defmodule Nx do
     impl!(tensor).sort(tensor, tensor, axis: axis, comparator: comparator)
   end
 
+  @doc """
+  Solve the equation a x = b for x, assuming a is a triangular matrix.
+
+  ## Examples
+
+      iex> Nx.triangular_solve([[3, 0, 0, 0], [2, 1, 0, 0], [1, 0, 1, 0], [1, 1, 1, 1]], [4, 2, 4, 2])
+      #Nx.Tensor<
+        f64[4]
+        [1.33333333, -0.66666667, 2.66666667, -1.33333333]
+      >
+
+  ### Error cases
+
+      iex> Nx.triangular_solve([[3, 0, 0, 0], [2, 1, 0, 0], [1, 0, 1, 0], [1, 1, 1, 1]], [4, 2, 4, 2], trans: 9)
+      ** (ArgumentError) unknown trans 9, expected 0 or 'N' for a x = b
+
+      iex> Nx.triangular_solve([[3, 0, 0, 0], [2, 1, 0, 0]], [4, 2, 4, 2], trans: 0)
+      ** (ArgumentError) expected a square matrix, got: {2, 4}
+
+      iex> Nx.triangular_solve([[3, 0, 0, 0], [2, 1, 0, 0], [1, 1, 1, 1], [1, 1, 1, 1]], [4], trans: 0)
+      ** (ArgumentError) incompatible dimensions
+  """
+  @doc type: :linalg
+  def triangular_solve(a, b, opts \\ []) do
+    %T{shape: s1} = a = tensor(a)
+    %T{shape: s2} = b = tensor(b)
+
+    case shape(s1) do
+      {n, n} -> {n, n}
+      other -> raise ArgumentError, "expected a square matrix, got: #{inspect(other)}"
+    end
+
+    if size(a) / 4 != size(b), do: raise(ArgumentError, "incompatible dimensions")
+
+    assert_keys!(opts, [:trans])
+
+    case opts[:trans] do
+      0 ->
+        triangular_solve_n(a, b)
+
+      'N' ->
+        triangular_solve_n(a, b)
+
+      nil ->
+        triangular_solve_n(a, b)
+
+      trans ->
+        raise ArgumentError, "unknown trans #{inspect(trans)}, expected 0 or 'N' for a x = b"
+    end
+  end
+
+  defp triangular_solve_n(a, b) do
+    # TODO: write implementation
```

Comment:
@AlexJuca I wrote a draft [here](https://github.com/elixir-nx/nx/issues/268#issuecomment-783352908) that might be of interest to you. My draft had an upper triangular matrix in mind, but with some minor changes I think the idea remains the same for lower triangular ones

Reply:
@polvalente Hey thanks, I will have a look at it and by end of the day I will have an implementation.

Reply:
@josevalim Does this particular implementation need to be done via the binary backend?

Reply:
Yes! The QR decomposition is a good example.

Reply:
> Yes! The QR decomposition is a good example.

Thanks